### PR TITLE
Adding remaining support for math types

### PR DESCRIPTION
**Pull request by scottenglert**
I am adding the remaining support for MVector, MPoint, etc that is missing in the math bindings.

### CODE REVIEW

**Comment by mottosso** [comment]
Yes, there are a few topics worthy of a discussion in your other pr #29 but I'd imagine it'd be better to make progress and revise later than to hold things up indefinitely. Primarily, the idea of diverging from the official API into a more user-friendly variant. I think that's a good idea and there's a lot of room for experimentation there. What I've got going with cmdx has been rounded off over the past few years and is something I think could work well here as well, so that would be my preferred starting point.

About merge permissions, I honestly can't find it. 😅 I think they've changed the way permissions work here, but you are already a "collaborator" and it's not clear to me whether you are also able to merge? If not, just give me a sign and I'll hit the merge button when you're ready. Don't worry about straying too far off into experimentation land, there's always git history. If you tag commits containing some big decision that could help future review.

**Comment by scottenglert** [comment]
I think you already gave me merge permissions. I see it in this PR to merge.

**Comment by mottosso** [comment]
Great, then feel free to merge when happy. 🥳

File: src/MFnTransform.inl
Comment (Muream) [inline src/MFnTransform.inl:173]: In order to fix the stubs generation you'll need to use `py::arg_v` instead of `py::arg`.
This is because the `parent` argument has a default value that doesn't map well to python.
`py::arg_v` lets you provide the default value as well as the string that should be used in the stubs for that default value.

There's a bit more information on that [here](https://github.com/mottosso/cmdc/blob/master/.github/contributing.md#ensuring-good-stubs-generation) as well as links to pybind's documentation that goes a bit more in depth.


Here's what this line should be:
```suggestion
    }, py::arg_v("parent", MObject::kNullObj, "Object.kNullObj"), _doc_FnTransform_create)
```

File: src/MTransformationMatrix.inl
Comment (Muream) [inline src/MTransformationMatrix.inl:85]: You should use `py::arg` in order for the stubs to get the correct name and value for the `interp` argument.

```suggestion
    }, py::arg("interp") = 1.0, _doc_TransformationMatrix_asMatrix)
```

File: src/MTransformationMatrix.inl
Comment (Muream) [inline src/MTransformationMatrix.inl:170]: You should add `py::arg` here as well.
I've named the argument `q` here to match Maya's documentation though I'm not super fond of these short names.

```suggestion
    .def("setRotationOrientation", &MTransformationMatrix::setRotationOrientation, py::arg("q"), _doc_TransformationMatrix_setRotationOrientation)
```

File: src/Math.inl
Comment (Muream) [inline src/Math.inl:20]: And you need `py::arg` here as well. Again, I've kept the same argument name as in maya's doc.
```suggestion
    }), py::arg("d"), "Create a new vector from a seqence or 2 or 3 floats")
```

**Review by Muream — CHANGES_REQUESTED** [review]
I noticed the stubs generation was blocking this, I've suggested a few changes to fix that.

I'll push a PR soon to make the stubs generation errors a bit clearer as well

**Comment by scottenglert** [comment]
Thanks! I might have forgot a couple of the arguments to document, it's easy to do if you are copying and pasting. Thank you for explaining to me the different between py::arg and py::arg_v with the stub generation and reread that page again. I'll incorporate your changes and be more watchful on this going forward. The documenation for pybind is not as comprehensive as I hoped :-)

**Comment by Muream** [comment]
Yeah, it took me a bit to understand how to generate nice stubs when I set that up haha

If you ever run into any issues with them feel absolutely free to give me a ping, it's been a while since I've set that up but I can probably help anyway 👍

**Comment by mottosso** [comment]
This is starting to look like something that could replace my use of cmdx! Let me throw in some loose thoughts on how to make working with nodes more friendly than passing MObjects around.

1. Provide 2 objects, that more or less wraps MFnDependencyNode and MFnDagNode
2. Keep them separate from the API itself, like an API ontop of the API, that calls on the underlying API

That way, you can use them, or not. They would have no effect on the overall API we're wrapping.

```py
cube = cmdc.node("pCube1")  # A `DagNode` instance, provides everything from `MFnDependencyNode` and `MFnDagNode`
time = cmdc.node("time1") # A `Node` instance, only provides functionality from `MFnDependencyNode`
```

Whereby `cmdc.node` is a factory function, something that takes the path (or MObject) to a node and spits out the corresponding class for it. Initially, I expect we'll only ever need Node and DagNode - as opposed to a unique class for every node type in Maya (like PyMEL). But even in cmdx, there are a handful of exceptions. For example `ObjectSet` provides an `append` method. But those could easily be free functions as well, `cmdc.append(objset, cube)` which in retrospect I would have preferred.

Additionally, I expect we'll find the features of these function sets gets us a long way to providing an interface to a node, but there will be things we'll want to add that isn't strictly related to those. One of those being attribute access.

```py
attr = cube["translateY"]  # A `Plug` instance
```

This being another wrapper around a function set-like class, the MPlug. I personally like the syntactical separation between `.methods()` and `["attribute"]` access and it has worked well so far with cmdx, but alternatively it could be one option of getting to the attribute as it's subjective and syntax sugar only.

```py
attr = cube.findPlug("translateY")
```

The next big ticket item is undo. With cmdx, it would have been nigh impossible and impractical to use the Maya API without undo, but it is a huge and surprisingly challenging task. The way I've written it for cmdx is portable and should hold up better in C++ so I'd suggest we start there. It's implemented as a command plug-in.

Which brings me to where undo comes into play in practice, the `MDagModifier`. In cmdx, this is wrapped as natively as possible, with one exception.

```py
with cmdx.DagModifier() as mod:
  mod.create_node("transform")  # Undoable
```

Any use as a context manager comes with built-in undo/redo. It's optional.

```py
mod = cmdx.DagModifier(undoable=True)
with cmdx.DGModifier(undoable=False) as mod:
  ...
```

And that's about it. With these 4 concepts riding on-top of the underlying API, I expect we'd have a user-friendly "API to the API" with a reasonable amount of effort and without sacrificing the purity of a plain wrapper. Meaning it could still be used as-is, in plug-ins and without these "convenience wrappers" which I expect will be appreciated, especially if converting from C++ examples or just transitioning into cmdc.

**Comment by scottenglert** [comment]
I do like the idea of having a more user friendly layer on top of the API that makes working with these types more intuitive. Getting the correct MPlug that is composed of array and compounds takes several lines of code and if my memory serves me right, doesn't work via MFNDependencyNode.findPlug. So anything we can do to embed these API pathways into a more direct method will definitely be more readable and less error prone.

For this PR, I think I have all the functions for the Quaternion, Point, Vector, and Matrix that match up to the API 2.0 closely.  I haven't finished all the test cases yet, I will try to do that so at least each method is called. Then we can merge this in.

The last thing I was thinking since it was supported in pybind11 was the support for numpy datatypes. That would possibly make major speed improvements if for example you had an array of points and you want to multiply it by a matrix. Skip the Python intermediate step and give them the data ready to go. 
I'm not sure exactly how to add this as I haven't looked into the requirements, but seems doable.

**Comment by mottosso** [comment]
> Skip the Python intermediate step and give them the data ready to go.

That would be interesting! Another idea I've toyed with is getting Taichi in there.

- https://github.com/taichi-dev/taichi

With numpy to handle the data pipeline between Maya and Python, I would not be surprised if we could see near-native speeds from something like this with Python plug-ins in Maya.

**Pull request merged**
